### ISSUE TITLE
"$ python manage.py db init" and "migrate" commands doesn't run as expected

### ISSUE THREAD

**Issue by Littelarch** [miguelgrinberg/Flask-Migrate#173]
Hello. I'm learning to use flask-migrate but meet a problem when trying it on my new project.

I'm trying to use flask-migrate with flask_script and here is the manage.py file:

```
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys

from flask_sqlalchemy import SQLAlchemy
from flask_script import Manager
from flask_script import Command
from flask_script import prompt_bool
from flask_migrate import Migrate, MigrateCommand

from web.app import create_app
from web.models import drop_all

app = create_app()
db = SQLAlchemy(app)
migrate = Migrate(app, db)
manager = Manager(app)
manager.add_command('db', MigrateCommand)


class DropDataBaseTables(Command):
    def __init__(self):
        super(DropDataBaseTables, self).__init__()

    def run(self):
        """Drop all tables after user has verified."""
        verified = prompt_bool(
            'Do you really want to drop all the database tables?')
        if verified:
            drop_all()
            sys.stdout.write('All tables droped. Database is now empty.\n')


manager.add_command('drop_db', DropDataBaseTables())

if __name__ == '__main__':
    manager.run()
```

And then I followed the tutorial to run 'init' and 'migrate' command.
After the 'init' command, the terminal shows:

```
> (env_ebsproxy) arch_learning3@arch-VirtualBox2:ebsproxy (develop*)$ python manage.py db init
> /home/arch_learning3/workplace/ebs_gitlab/ebsproxy/env_ebsproxy/local/lib/python2.7/site-packages/flask_sqlalchemy/__init__.py:839: FSADeprecationWarning: SQLALCHEMY_TRACK_MODIFICATIONS adds significant overhead and will be disabled by default in the future.  Set it to True or False to suppress this warning.
>   'SQLALCHEMY_TRACK_MODIFICATIONS adds significant overhead and '
>   Creating directory /home/arch_learning3/workplace/ebs_gitlab/ebsproxy/ebsproxy/migrations ... done
>   Creating directory /home/arch_learning3/workplace/ebs_gitlab/ebsproxy/ebsproxy/migrations/versions ... done
>   Generating /home/arch_learning3/workplace/ebs_gitlab/ebsproxy/ebsproxy/migrations/env.py ... done
>   Generating /home/arch_learning3/workplace/ebs_gitlab/ebsproxy/ebsproxy/migrations/README ... done
>   Generating /home/arch_learning3/workplace/ebs_gitlab/ebsproxy/ebsproxy/migrations/script.py.mako ... done
>   Generating /home/arch_learning3/workplace/ebs_gitlab/ebsproxy/ebsproxy/migrations/alembic.ini ... done
>   Generating /home/arch_learning3/workplace/ebs_gitlab/ebsproxy/ebsproxy/migrations/env.pyc ... done
>   Please edit configuration/connection/logging settings in
>   '/home/arch_learning3/workplace/ebs_gitlab/ebsproxy/ebsproxy/migrations/alembic.ini' before proceeding.
```

and I check mysql tables:

```
mysql> show tables;
+---------------------+
| Tables_in_ebs_proxy |
+---------------------+
| alembic_version     |
| application         |
| course_proxy        |
| user_proxy          |
+---------------------+
4 rows in set (0.00 sec)

```

then I run 'migrate' command, the terminal shows:
```
(env_ebsproxy) arch_learning3@arch-VirtualBox2:ebsproxy (develop*)$ python manage.py db migrate
/home/arch_learning3/workplace/ebs_gitlab/ebsproxy/env_ebsproxy/local/lib/python2.7/site-packages/flask_sqlalchemy/__init__.py:839: FSADeprecationWarning: SQLALCHEMY_TRACK_MODIFICATIONS adds significant overhead and will be disabled by default in the future.  Set it to True or False to suppress this warning.
  'SQLALCHEMY_TRACK_MODIFICATIONS adds significant overhead and '
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.autogenerate.compare] Detected removed table u'user_proxy'
INFO  [alembic.autogenerate.compare] Detected removed table u'application'
INFO  [alembic.autogenerate.compare] Detected removed table u'course_proxy'
  Generating /home/arch_learning3/workplace/ebs_gitlab/ebsproxy/ebsproxy/migrations/versions/b6052e78976d_.py ... done
```

Why dose it create table when 'init' command and detected "removed table" when first run 'migrate'? Do I miss anything?

**Comment by miguelgrinberg** [comment]
The `init` command only creates the `alembic_version` table, which is used internally by Alembic (a dependency of Flask-Migrate). You seem to have three more tables in your database. How did those get there? Maybe you created them before you started using Flask-Migrate?

The `migrate` command looks at the state of the database and the state of the models in your application, and then generates a migration script that makes the database match the models. For that to work you need to make sure the models are imported, my guess is that the tables are getting deleted because they are not imported when the `migrate` command runs.

So basically you have two things to resolve:

- Delete the tables from your database, so that Flask-Migrate can create them through migrations.
- Import your models in your `manage.py` file, so that they are visible to the migration framework.

**Comment by Littelarch** [comment]
@miguelgrinberg Thanks for your reply.

As for the `init` problem, I only created a new database with out any table before ran 'init' command. 

But actually there is an initial method like `"Base.metadata.create_all(bind=engine)"` called in `"create_app()"`(the Line 15 in "manage.py" file). Should I remove this initial method if I use flask-migration?

----------------------------------------------------------------------------------------------------------------
As for the `migrate `problem, I tried to import all models into "manage.py" by `from web.models import application, course_proxy, user_proxy` but it still the same problem.

But I noticed something different between my project and tutorial. The model in my project is not defined by inheriting `db.Model`. Instead, there is an `BaseModel` defined like:
```
……(some other codes)

@as_declarative()
class BaseModel(object):
    """Base class used for database model. It adds adds fields to all model
    classes that subclass it.
    """

    @declared_attr
    def __tablename__(self):
        name = self.__name__

……(some other codes)
```
And other models are defined by:
```
class Application(BaseModel):
    """Implements the Application model."""

    name = Column(Unicode(255), unique=True, index=True)
    title = Column(Unicode(255))
    secrete = Column(Unicode(255))

    def __init__(self, name, title):
        self.name = name
        self.title = title
……
```

Does flask-migrate only work with the models defined by inhering  `db.Model`?

P.S. Sorry for not mentioned but this project is not created by me. I'm still not very familiar with it's code. I just try to add migration management on it.

**Comment by miguelgrinberg** [comment]
> Should I remove this initial method if I use flask-migration?

Yes. When you work with a migration framework you have to let the framework create and maintain your tables. Any other code that deals with table creation/deletion/modification should be removed.

Regarding your `migrate` problem, this extension is designed to work with Flask-SQLAlchemy, not with raw SQLAlchemy. Not sure what kind of hybrid do you have, because I see in your code that you do have Flask-SQLAlchemy. Maybe the problem comes from that, I really have no way to tell, but my guess is that the models are all loaded into a pure SQLAlchemy metadata object, while the metadata object maintained by Flask-SQLAlchemy is empty. I suggest you move your models to Flask-SQLAlchemy. If that is not a possibility, then I think you will not be able to use this extension, and instead may want to consider working with Alembic directly.

**Comment by Littelarch** [comment]
Understand. Thanks for your suggestion. It's very helpful!

**Comment by miguelgrinberg** [comment]
This issue will be automatically closed due to being inactive for more than six months. Seeing that I haven't responded to your last comment, it is quite possible that I have dropped the ball on this issue and I apologize about that. If that is the case, do not take the closing of the issue personally as it is an automated process doing it, just reopen it and I'll get back to you.
